Fix `@simd` for non 1 step CartesianPartition

N5N3 opened the pull request:
Previous code assumes that the steps of each axes are 1, which is not valid after #37829.
This PR tries to add support for non 1 step `CartesianPartition`.
Before
```julia
julia> @simd for i in view(CartesianIndices((1:2:3,1:2:3)),1:4)
       println(i)
       end
CartesianIndex(1, 1) # wrong result
CartesianIndex(2, 1)
CartesianIndex(3, 1)
CartesianIndex(1, 2)
CartesianIndex(2, 2)
CartesianIndex(3, 2)
CartesianIndex(1, 3)
CartesianIndex(2, 3)
CartesianIndex(3, 3)
julia> @simd for i in view(CartesianIndices((1:2:3,1:2:3,1:2:4)),1:4)
       println(i)
       end
ERROR: MethodError: no method matching LinearIndices(::Tuple{StepRange{Int64, Int64}, StepRange{Int64, Int64}})
```
After
```julia
julia> @simd for i in view(CartesianIndices((1:2:3,1:2:3)),1:4)
       println(i)
       end
CartesianIndex(1, 1)
CartesianIndex(3, 1)
CartesianIndex(1, 3)
CartesianIndex(3, 3)
julia> @simd for i in view(CartesianIndices((1:2:3,1:2:3,1:2:4)),1:4)
       println(i)
       end
CartesianIndex(1, 1, 1)
CartesianIndex(3, 1, 1)
CartesianIndex(1, 3, 1)
CartesianIndex(3, 3, 1)
```

Besides the above fix, this PR also tries to simplify the inner loop with a `Generator` based outer range. It accelerates 3/4d CartesianPartition a little on my desktop.
Some benchmarks:
```julia
julia> f(a) = @inbounds @simd for i in view(CartesianIndices(a),2:length(a)-1)
       a[i] = 1.0
       end
julia> a = zeros(256,256);
julia> @btime f($a)
  10.700 μs (0 allocations: 0 bytes)  # before: 10.900 μs (0 allocations: 0 bytes)  
julia> a = zeros(64,64,64);
julia> @btime f($a)
  44.500 μs (0 allocations: 0 bytes)  # 49.700 μs (0 allocations: 0 bytes)
julia> a = zeros(16,16,16,16);
julia> @btime f($a)
  14.000 μs (0 allocations: 0 bytes)  # 15.800 μs (0 allocations: 0 bytes)
```

N5N3 commented:
Although strange, the performance of 4D `@simd` does increase after modification.
Now the `@simd` performance of `CartesianPartition` is close to `CartesianIndices`.
Benchmark code:
```julia
bench1(f::OP, a, b = ()) where OP = begin
    @inbounds @simd for i in eachindex(IndexCartesian(), a)
        a[i] = f(map(x -> (@inbounds x[i]), b)...)
    end
end
bench2(f::OP, a, b = ()) where OP = begin
    iter = view(eachindex(IndexCartesian(), a), 2:length(a) - 1)
    @inbounds @simd for i in iter
        a[i] = f(map(x -> (@inbounds x[i]), b)...)
    end
end
println("----------fill!------------")
N = 2^16
for n = 2:4
    a = zeros(N>>(3n-3), ntuple(_ -> 8, n - 1)...)
    println("N = ", n)
    @btime bench1(Returns(1), $a)
    @btime bench2(Returns(1), $a)
end
println("---------- .* ------------")
N = 2^16
for n = 2:4
    a = zeros(N>>(3n-3), ntuple(_ -> 8, n - 1)...)
    b = randn(size(a))
    c = randn(size(a))
    println("N = ", n)
    @btime bench1(*, $a, ($c, $b))
    @btime bench2(*, $a, ($c, $b))
end
```
Result:
```julia
----------fill!------------
N = 2
  10.800 μs (0 allocations: 0 bytes)
  11.100 μs (0 allocations: 0 bytes)
N = 3
  10.700 μs (0 allocations: 0 bytes)
  10.900 μs (0 allocations: 0 bytes)
N = 4
  11.100 μs (0 allocations: 0 bytes)
  10.800 μs (0 allocations: 0 bytes)
---------- .* ------------
N = 2
  22.700 μs (0 allocations: 0 bytes)
  22.800 μs (0 allocations: 0 bytes)
N = 3
  22.700 μs (0 allocations: 0 bytes)
  22.800 μs (0 allocations: 0 bytes)
N = 4
  22.800 μs (0 allocations: 0 bytes)
  23.300 μs (0 allocations: 0 bytes)
```

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/42736#issuecomment-949798349) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/3dfd4b6_vs_59aa3ed/report.md).

N5N3 committed:
drop support of view(ReshapedArrayLF)

revert changes in reshapedarray.jl
use Iterators.rest

N5N3 commented on base/multidimensional.jl:
```diff
@@ -616,33 +615,45 @@ module IteratorsMD
+        vr, fr = divrem(last(iter.indices[1]) - 1, mi[end]) .+ (1, first(ax1))
+        oci = CartesianIndices(tail(ci.indices))
+        # A fake CartesianPartition to reuse the outer iterate fallback
+        outer = @inbounds view(ReshapedArray(oci, (length(oci),), mi), vl:vr)
```
All test passed except for freebsd64, so I think this PR is ready for further review. @vtjnash @mbauman
Since `ReshapedArray` use stride based `mi` to accelerate index transformation, I have to form the `simd_outer_range` with an illegal `CartesianPartition`. If `mi` carries size infomation, then we can reuse the input's `mi` safely, and there's no need to use `Iterators.rest`. 

Previous commit shows that official test passed with size based `mi` (but pkgs might not). My local bench shows that size based `mi` could accelerate the sequential access of a ReshapedArray a little.
@timholy Is there a strong reason for us to use stride based `mi`? And is it acceptable to make `mi` carry size infomation in this PR?

vchuravy commented:
@nanosoldier `runbenchmarks("simd", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/42736#issuecomment-955762656) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/3522071_vs_86f656d/report.md).

N5N3 commented:
Interesting results.
The first possible regression also appeared in [Nanosoldier Daily Job: 2021-10-28 vs 2021-10-26](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_date/2021-10/28/report.md) with a close time ratio (1.77 there and 1.76 here).
As for the second (and third, they are exactly the same), similar `possibe regression` on `("axpy!", "Float32", 4096)` could also be found in the above report.
So I think we can ingore them here.

BTW, is there any work remain to do on https://github.com/JuliaCI/BaseBenchmarks.jl/pull/284 @vchuravy

N5N3 committed:
Update broadcast.jl

move `@inbounds` outside the loop body. see #38086

N5N3 commented:
Just notice that reused `simd_index` makes the position of `@inbounds` influence `@simd for`'s performance. (`BitArray`'s `broadcast` is the only use case I found in Base.)

On the other hand, I still think we'd better make `CartesianIndices`'s `simd_index` always inbound like before, since it is really strange that `@inbounds` influences the performance of index iteration.

N5N3 commented:
Rebasing solves Downloads error on FreeBSD.
The current `threads` error on Linux32 seems unrelated.
Just wonder whether further test is needed?
Or just re-run simd `Benchmark`? @vchuravy  @vtjnash

N5N3 committed:
use `firstindex` in `CartesianIndices`'s `simd_index`

`first(Base.axes1())` works well, but `length(::StepRange)` won't be omitted by LLVM if `step` is unknown at compile time.

N5N3 commented:
Bump?

KristofferC commented:
Rebased to get another CI run.

N5N3 commented:
Thanks for pick this up.
Error on "julia-master" should be https://github.com/JuliaLang/julia/issues/42752 (thus unrelated).
